akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala

```diff
@@ -218,8 +277,18 @@ private[http2] abstract class Http2Demux(http2Settings: Http2CommonSettings, ini
             // are enforced from the start of the connection so there's no need to invoke
             // `enforceSettings(initialLocalSettings)`
 
-            case PingFrame(true, _)  =>
-            // ignore for now (we don't send any pings)
+            case PingFrame(true, _) =>
+              pingState match {
+                case OptionVal.Some(state) =>
+                  state.pingsWithoutData += 1L
+                  if (settings.maxPingsWithoutData > 0) {
+                    // FIXME fail with specific exception or some more graceful failure?
+                    if (state.pingsWithoutData > settings.maxPingsWithoutData)
+                      throw new RuntimeException("HTTP/2 more than " + settings.maxPingsWithoutData + " pings exchanged without any data frames. This is configureable by akka.http2.[client, server].max-keepalives-without-data")
+                  }
+                // FIXME also fail if too much time passed since last ping was sent?
```

Comment:
Is this about the race condition between receiving the ack and running `onTimer`? I think no need to care for that, after all we received an ack. (Just need to make sure that `onTimer` doesn't kill the connection afer we handled the ack here, I cannot see if that's done currently?)

Reply:
I think maybe that race isn't very interesting and we can keep it only failing on timer if too much time passed without seeing ack.

Reply:
Agreed.